File: MedianStream.java
class Solution {
    PriorityQueue<Integer> max = new PriorityQueue<>(Collections.reverseOrder());
    PriorityQueue<Integer> min = new PriorityQueue<>();
    public ArrayList<Double> getMedian(int[] arr) {
        // code here
        ArrayList<Double> res = new ArrayList<>();
        for(int ele : arr){
            insertInHeap(ele);
            balance();
            res.add(getMedian());
        }
        return res;
    }
    
    double getMedian(){
        if(max.size() == min.size()){
            return (double)(max.peek() + min.peek()) / 2;
        }
        else{
            return max.peek();
        }
    }
    
    void balance(){
        if(max.size() > min.size() + 1){
            min.add(max.poll());
        }
        else if(max.size() < min.size()){
            max.add(min.poll());
        }
    }
    
    void insertInHeap(int ele){
        if(max.isEmpty() || max.peek() > ele){
            max.add(ele);
        }
        else{
            min.add(ele);
        }
    }
    
}

File: MedianStream.cpp
class Solution {
priority_queue<int> maxHeap; // Max heap (stores the smaller half)
    priority_queue<int, vector<int>, greater<int>> minHeap; // Min heap (stores the larger half)

public:
    vector<double> getMedian(vector<int>& arr) {
        vector<double> medians;
        for (int ele : arr) {
            insertInHeap(ele);
            balanceHeap();
            medians.push_back(getMedian());
        }
        return medians;
    }

    double getMedian() {
        if (maxHeap.size() == minHeap.size()) {
            return (maxHeap.top() + minHeap.top()) / 2.0;
        } else {
            return maxHeap.top();
        }
    }

    void balanceHeap() {
        if (maxHeap.size() > minHeap.size() + 1) {
            minHeap.push(maxHeap.top());
            maxHeap.pop();
        } else if (maxHeap.size() < minHeap.size()) {
            maxHeap.push(minHeap.top());
            minHeap.pop();
        }
    }

    void insertInHeap(int ele) {
        if (maxHeap.empty() || maxHeap.top() > ele) {
            maxHeap.push(ele);
        } else {
            minHeap.push(ele);
        }
    }
};
